### PR TITLE
Patch Property Definition in terraform for new changes

### PR DESCRIPTION
**Pull request by taimoor-at-opslevel**
## Issues

https://github.com/OpsLevel/team-platform/issues/160

## Changelog

- [x] Re-add resource, add new description and visibility field
- [x] Add update function in resource
- [x] Re-add datasources
- [x] Make a `changie` entry

### Create a property definition (only provide name and schema, property_display_status will use default)

```
  # opslevel_property_definition.newnewpd will be created
  + resource "opslevel_property_definition" "newnewpd" {
      + id                      = (known after apply)
      + last_updated            = (known after apply)
      + name                    = "new pd tf"
      + property_display_status = "visible"
      + schema                  = jsonencode(
            {
              + "$defs" = {
                  + MyProp = {
                      + additionalProperties = false
                      + properties           = {
                          + name = {
                              + default     = "alex"
                              + description = "The name of a friend"
                              + examples    = [
                                  + "joe",
                                  + "lucy",
                                ]
                              + title       = "the name"
                              + type        = "string"
                            }
                        }
                      + required             = [
                          + "name",
                        ]
                      + type                 = "object"
                    }
                }
              + "$ref"  = "#/$defs/MyProp"
            }
        )
    }

Plan: 1 to add, 0 to change, 0 to destroy.
```

### Add a description, change visibility (keep name and schema unchanged)

```
  # opslevel_property_definition.newnewpd will be updated in-place
  ~ resource "opslevel_property_definition" "newnewpd" {
      + description             = "hello world - this was updated"
        id                      = "Z2lkOi8vb3BzbGV2ZWwvUHJvcGVydGllczo6RGVmaW5pdGlvbi8xNjU"
        name                    = "new pd tf"
      ~ property_display_status = "visible" -> "hidden"
        # (1 unchanged attribute hidden)
    }

Plan: 0 to add, 1 to change, 0 to destroy.
```

### Change the schema, visibility (keep name and description unchanged)

```
  # opslevel_property_definition.newnewpd will be updated in-place
  ~ resource "opslevel_property_definition" "newnewpd" {
        id                      = "Z2lkOi8vb3BzbGV2ZWwvUHJvcGVydGllczo6RGVmaW5pdGlvbi8xNjU"
        name                    = "new pd tf"
      ~ property_display_status = "hidden" -> "visible"
      ~ schema                  = jsonencode(
          ~ {
              - "$defs" = {
                  - MyProp = {
                      - additionalProperties = false
                      - properties           = {
                          - name = {
                              - default     = "alex"
                              - description = "The name of a friend"
                              - examples    = [
                                  - "joe",
                                  - "lucy",
                                ]
                              - title       = "the name"
                              - type        = "string"
                            }
                        }
                      - required             = [
                          - "name",
                        ]
                      - type                 = "object"
                    }
                }
              - "$ref"  = "#/$defs/MyProp"
              + enum    = [
                  + "red",
                  + "green",
                  + "blue",
                ]
              + type    = "string"
            }
        )
        # (2 unchanged attributes hidden)
    }

Plan: 0 to add, 1 to change, 0 to destroy.
```

### Delete the property definition

```
  # opslevel_property_definition.newnewpd will be destroyed
  - resource "opslevel_property_definition" "newnewpd" {
      - description             = "hello world - this was updated" -> null
      - id                      = "Z2lkOi8vb3BzbGV2ZWwvUHJvcGVydGllczo6RGVmaW5pdGlvbi8xNjU" -> null
      - last_updated            = "Thursday, 21-Dec-23 12:52:55 EST" -> null
      - name                    = "new pd tf" -> null
      - property_display_status = "visible" -> null
      - schema                  = jsonencode(
            {
              - enum = [
                  - "red",
                  - "green",
                  - "blue",
                ]
              - type = "string"
            }
        ) -> null
    }

Plan: 0 to add, 0 to change, 1 to destroy.
```

### Test datasource after creating a resource

main.tf:

```
resource "opslevel_property_definition" "newnewpd" {
    name = "new pd tf"
    description = "hello world - this was updated"
    schema = jsonencode({"enum": ["red","green","blue"],"type": "string"})
    # property_display_status = "hidden"
}

data "opslevel_property_definition" "load_newnewpd" {
  identifier = opslevel_property_definition.newnewpd.id
}

output "out_newnewpd" {
  value = data.opslevel_property_definition.load_newnewpd
}
```

```
out_newnewpd = {
  "id" = "Z2lkOi8vb3BzbGV2ZWwvUHJvcGVydGllczo6RGVmaW5pdGlvbi8xNjY"
  "identifier" = "Z2lkOi8vb3BzbGV2ZWwvUHJvcGVydGllczo6RGVmaW5pdGlvbi8xNjY"
  "name" = "new pd tf"
  "schema" = "{\"enum\":[\"red\",\"green\",\"blue\"],\"type\":\"string\"}"
}
```

### Test plural datasource

main.tf:

```
data "opslevel_property_definitions" "allpds" {}

output "out_allpds" {
  value = data.opslevel_property_definitions.allpds
}
```

```
out_allpds = {
  "descriptions" = tolist([
    "hello David",
    "",
    "delete this",
    "esefsefwa",
    "esefsefwa",
    "Test - delete it",
    "Test - delete it",
    "JSON input acting funny",
    "JSON input acting funny",
    "Test - delete it",
    "",
    "",
  ])
  "id" = "1703182164"
  "ids" = tolist([
    "Z2lkOi8vb3BzbGV2ZWwvUHJvcGVydGllczo6RGVmaW5pdGlvbi8xNjI",
    "Z2lkOi8vb3BzbGV2ZWwvUHJvcGVydGllczo6RGVmaW5pdGlvbi85OA",
    "Z2lkOi8vb3BzbGV2ZWwvUHJvcGVydGllczo6RGVmaW5pdGlvbi8xNTE",
    "Z2lkOi8vb3BzbGV2ZWwvUHJvcGVydGllczo6RGVmaW5pdGlvbi82OA",
    "Z2lkOi8vb3BzbGV2ZWwvUHJvcGVydGllczo6RGVmaW5pdGlvbi82OQ",
    "Z2lkOi8vb3BzbGV2ZWwvUHJvcGVydGllczo6RGVmaW5pdGlvbi8xNTU",
    "Z2lkOi8vb3BzbGV2ZWwvUHJvcGVydGllczo6RGVmaW5pdGlvbi8xNTc",
    "Z2lkOi8vb3BzbGV2ZWwvUHJvcGVydGllczo6RGVmaW5pdGlvbi8xMTQ",
    "Z2lkOi8vb3BzbGV2ZWwvUHJvcGVydGllczo6RGVmaW5pdGlvbi8xNDI",
    "Z2lkOi8vb3BzbGV2ZWwvUHJvcGVydGllczo6RGVmaW5pdGlvbi8xNTM",
    "Z2lkOi8vb3BzbGV2ZWwvUHJvcGVydGllczo6RGVmaW5pdGlvbi82NQ",
    "Z2lkOi8vb3BzbGV2ZWwvUHJvcGVydGllczo6RGVmaW5pdGlvbi82Ng",
  ])
  "names" = tolist([
    "graphiql_updated",
    "Is Beta Feature",
    "pd",
    "Some Definition",
    "Some Definition (Clooooone)",
    "Test NULL",
    "Test NULL1223",
    "Test PropDef - delete me",
    "Test PropDef - delete me2",
    "Test PropDef - int only",
    "Test1",
    "Test2",
  ])
  "property_display_statuses" = tolist([
    "hidden",
    "visible",
    "visible",
    "visible",
    "visible",
    "visible",
    "visible",
    "visible",
    "visible",
    "visible",
    "visible",
    "visible",
  ])
  "schemas" = tolist([
    "{\"enum\":[\"red\",\"green\",\"blue\"],\"type\":\"string\"}",
    "{\"$schema\":\"https://json-schema.org/draft/2020-12/schema\",\"type\":\"boolean\"}",
    "{\"enum\":[\"red\",\"green\",\"blue\"],\"type\":\"string\"}",
    "{\"type\":\"boolean\"}",
    "{\"type\":\"boolean\"}",
    "{\"hey\":\"hi\"}",
    "{\"hey\":\"hi\"}",
    "{\"$schema\":\"https://json-schema.org/draft/2020-12/schema\",\"type\":\"boolean\"}",
    "{\"$schema\":\"https://json-schema.org/draft/2020-12/schema\",\"type\":\"boolean\"}",
    "{\"key\":\"value\"}",
    "{\"type\":\"string\"}",
    "{\"type\":\"string\"}",
  ])
}
```

### CODE REVIEW

File: opslevel/datasource_opslevel_property_definition.go [hunk 1]
```diff
@@ -41,9 +53,15 @@ func datasourcePropertyDefinitionRead(d *schema.ResourceData, client *opslevel.C
 	if err := d.Set("name", resource.Name); err != nil {
 		return err
 	}
+	if err := d.Set("description", resource.Description); err != nil {
+		return err
+	}
 	if err := d.Set("schema", resource.Schema.ToJSON()); err != nil {
 		return err
 	}
+	if err := d.Set("property_display_status", string(resource.PropertyDisplayStatus)); err != nil {
```
Comment (davidbloss) [inline opslevel/datasource_opslevel_property_definition.go:62]:
```suggestion
	if err := d.Set("property_display_status", opslevel.PropertyDisplayStatusEnum(resource.PropertyDisplayStatus)); err != nil {
```

File: opslevel/resource_opslevel_property_definition.go [hunk 2]
```diff
@@ -23,23 +24,34 @@ func resourcePropertyDefinition() *schema.Resource {
...
 				Description: "The schema of the property definition.",
-				ForceNew:    true,
 				Required:    true,
 			},
+			"property_display_status": {
+				Type:        schema.TypeString,
+				Description: "The display status of a custom property on service pages. (Options: 'visible' or 'hidden')",
+				Default:     "visible",
+				Optional:    true,
+			},
```
Comment (davidbloss) [inline opslevel/resource_opslevel_property_definition.go:44]:
```suggestion
				ValidateFunc: validation.StringInSlice(opslevel.AllPropertyDisplayStatusEnum, false),
			},
```

File: opslevel/resource_opslevel_property_definition.go [hunk 3]
```diff
@@ -61,13 +74,37 @@ func resourcePropertyDefinitionRead(d *schema.ResourceData, client *opslevel.Cli
...
+func resourcePropertyDefinitionUpdate(d *schema.ResourceData, client *opslevel.Client) error {
+	id := d.Id()
+	input := opslevel.PropertyDefinitionInput{
+		Name:                  d.Get("name").(string),
+		Description:           d.Get("description").(string),
+		Schema:                opslevel.NewJSON(d.Get("schema").(string)),
+		PropertyDisplayStatus: opslevel.PropertyDisplayStatusEnum(d.Get("property_display_status").(string)),
+	}
+
+	_, err := client.UpdatePropertyDefinition(id, input)
```
Comment (rocktavious) [inline opslevel/resource_opslevel_property_definition.go:99]: this is a thing???  I thought definitions only had create and delete?

**Review by rocktavious — CHANGES_REQUESTED** [review]
I think we are missing the examples?

File: opslevel/resource_opslevel_property_definition.go (hunk 3 above)
Comment (taimoor-at-opslevel) [reply opslevel/resource_opslevel_property_definition.go:99]: @rocktavious This was added very recently! We also have CLI supporting `update` on property definitions now. https://github.com/OpsLevel/cli/pull/225

**Comment by taimoor-at-opslevel** [comment]
> I think we are missing the examples?

@rocktavious the examples already exist in the repository and can be found here: https://github.com/OpsLevel/terraform-provider-opslevel/tree/main/examples/resources/opslevel_property_definition

File: opslevel/resource_opslevel_property_definition.go (hunk 3 above)
Comment (rocktavious) [reply opslevel/resource_opslevel_property_definition.go:100]: 🎉

**Review by rocktavious — APPROVED** [review]
🚀

**Pull request merged**